### PR TITLE
feat: Add wasm support for `pages functions build`

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**
**What this PR solves**
Currently it is not possible to import `wasm` modules in either Pages
Functions or Pages Advanced Mode projects.

This commit caries out work to address the aforementioned issue by
enabling `wasm` module imports in `wrangler pages functions build`.
As a result, Pages users can now import their `wasm` modules withing
their Functions or `_worker.js` files, and `wrangler pages functions
build` will correctly bundle everything and output the expected result
file.

```
import hello from "./hello.wasm"

export async function onRequest() {
	const module = await WebAssembly.instantiate(hello);
	return new Response(module.exports.hello);
}
```

**How to test**
This feature is behind the `[--experimental-worker-bundle]` flag

`npx wrangler pages functions build <options> --experimental-worker-bundle=true`

**Associated docs issues/PR**
N/A

**Author has included the following, where applicable**

- [x] Tests
- [x] Changeset

**Reviewer has performed the following, where applicable**

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Fixes # [insert issue number].

### CODE REVIEW

**Comment by ruslantalpa** [comment]
@CarmenPopoviciu Is it possible to try this PR as suggested by the comment from github-actions bot?
I tried that but still get "no loader defined for .wasm" error when running `npx wrangler pages dev/publish ...`

Everything works as expected (.wasm correctly bundled) when using `wrangler dev/publish` commands on the exact same file

Am I misunderstanding the scope of this PR or is it just in progress?

Thank you

PS: I am trying to have a pages/function that uses a dependency that has a line similar to 
` import wasmBin from './some_file.wasm'`

**Comment by CarmenPopoviciu** [comment]
@ruslantalpa this PR is still WIP, but you should be able to test it by running `npx wrangler pages functions bundle --experimental-worker-bundle=true`. Again, this is WIP, and we will probably release these changes under the existing `pages functions build` cmd. Another thing you should note is that this command is not really an "open" API. At this moment it is `hidden` and meant for internal use only. We are however considering maybe opening it up to external users at some point in the future, but there is no roadmap for that yet.

File: packages/wrangler/src/__tests__/helpers/string-dynamic-values-matcher.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Sometimes, we might need to test strings that contain dynamic |
+ * random | generated data (such as file hashes, IDs, etc.).
+ *
+ * This helper function replaces the dynamic parts of such strings with
+ * static values, thus enabling us to properly test the contents of the
+ * string.
+ *
+ * see https://jestjs.io/docs/snapshot-testing#property-matchers
+ */
+export function replaceRandomWithConstantData(
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/helpers/string-dynamic-values-matcher.ts:13]: I'm open to better naming here, if applicable. same goes for the file name

File: packages/wrangler/src/pages/build.ts [hunk 2]
```diff
@@ -121,32 +134,110 @@ export const Handler = async ({
...
-			buildOutputDirectory,
-			nodeCompat,
-			routesOutputPath,
+
+	const workerScriptPath = resolvePath(buildOutputDirectory, "_worker.js");
+	const foundWorkerScript = existsSync(workerScriptPath);
+	let bundle: BundleResult | undefined = undefined;
+
+	// `wrangler pages functions build --experimental-worker-bundle=true`
+	if (
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/build.ts:148]: not sure if necessary. I ran this cmd randomly in all the wrong folders by accident so many times that I got so annoyed and wrote a nice error message. It might come in handy once the flag is removed tho ¯\_(ツ)_/¯

**Comment by ruslantalpa** [comment]
@CarmenPopoviciu I have tried the experimental bundle flag and it's working great locally (packaging wasm correctly just like the "worker" path), will try a live deploy soon
I have the example here in case it that helps https://github.com/subzerocloud/showcase/tree/main/cloudflare-postgresql-neon

Hope you merge/release this soon. Thank you for the info on how to run this.

File: packages/wrangler/src/pages/build.ts [hunk 3]
```diff
@@ -121,32 +134,110 @@ export const Handler = async ({
 		}
 	}
 
+	// for now let's restrict `outfile` to `_worker.bundle` if the
+	// experimentalWorkerBundle flag is set. Once we remove the flag
+	// this will no longer be an issue
+	outfile = experimentalWorkerBundle ? "_worker.bundle" : outfile;
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/build.ts:140]:
```suggestion
```

I wouldn't bother with this.

File: packages/wrangler/src/pages/build.ts (hunk 2 above)
Comment (GregBrimble) [reply packages/wrangler/src/pages/build.ts:148]: Love it!

File: packages/wrangler/src/pages/build.ts (hunk 3 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/build.ts:140]: the default `outfile` is `_worker.js` atm. If I change that to `_worker.bundle` it'll break crr behaviour. I remember saying don't want to break crr behaviour?

File: packages/wrangler/src/pages/build.ts [hunk 4]
```diff
@@ -121,32 +134,110 @@ export const Handler = async ({
...
+	const workerScriptPath = resolvePath(buildOutputDirectory, "_worker.js");
+	const foundWorkerScript = existsSync(workerScriptPath);
+	let bundle: BundleResult | undefined = undefined;
+
+	// `wrangler pages functions build --experimental-worker-bundle=true`
+	if (
+		!isInPagesCI &&
+		experimentalWorkerBundle &&
+		!foundWorkerScript &&
+		directory === "functions" &&
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/build.ts:152]:
```suggestion
		!foundWorkerScript &&
```

Any reason _not_ to throw this error in more general circumstances too? I can make it so I'm only calling this when I know a that either `functions/` or `_worker.js` exists.

File: packages/wrangler/src/pages/build.ts (hunk 3 above)
Comment (GregBrimble) [reply packages/wrangler/src/pages/build.ts:140]: I'll manually specify `--outfile=_worker.bundle`. I put it in a known location anyway.

File: fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { CHOCOLATE } from "../chocolate";
```
Comment (GregBrimble) [inline fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts:1]: What's this testing? Just confirming that regular bundling works fine?

File: .changeset/thirty-hornets-behave.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+---
+
+Currently it is not possible to import `wasm` modules in either Pages
+Functions or Pages Advanced Mode projects.
+
+This commit caries out work to address the aforementioned issue by
+enabling `wasm` module imports in `wrangler pages functions build`.
+As a result, Pages users can now import their `wasm` modules withing
+their Functions or `_worker.js` files, and `wrangler pages functions build` will correctly bundle everything and output the expected result
+file.
```
Comment (GregBrimble) [inline .changeset/thirty-hornets-behave.md:12]: I think Changesets should be less of a 'commit message', and more of a 'release note'.

I'd maybe rephrase to something like:

```suggestion
You can now import Wasm modules in Pages Functions and Pages Functions Advanced Mode (`_worker.js`).

This change specifically enables `wasm` module imports in `wrangler pages functions build`.
As a result, Pages users can now import their `wasm` modules within
their Functions or `_worker.js` files, and `wrangler pages functions build` will correctly bundle everything and output the expected result
file.
```

File: packages/wrangler/src/__tests__/helpers/string-dynamic-values-matcher.ts (hunk 1 above)
Comment (GregBrimble) [reply packages/wrangler/src/__tests__/helpers/string-dynamic-values-matcher.ts:13]: Fine by me! People can rename if they want to abstract it further when they re-use :)

**Review by GregBrimble — APPROVED** [review]

**Comment by GregBrimble** [comment]
Just FYI @ruslantalpa, there are some changes we need to make to our API backend before this will be available in production. You should be able to test this locally with `wrangler pages dev` already :)

File: fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts (hunk 5 above)
Comment (CarmenPopoviciu) [reply fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts:1]: yes, that's correct. Since we never bundled _worker.js before, I wanted to make sure we have that covered in a fixture

Comment (GregBrimble) [reply fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts:1]: Understood, but this is a `functions/` dir, yes? You don't have a fixture test checking non-Wasm imports in a `_worker.js`, (but you have covered it extensively in regular tests so absolutely not necessary imo!)

File: packages/wrangler/src/pages/build.ts (hunk 3 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/build.ts:140]: ✅

File: fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts (hunk 5 above)
Comment (CarmenPopoviciu) [reply fixtures/pages-functions-wasm-app/functions/other-meaning-of-life.ts:1]: lol...ignore me, I'm a pinhead 🙄

File: packages/wrangler/src/pages/build.ts (hunk 4 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/build.ts:152]: ✅

File: packages/wrangler/src/__tests__/pages/functions-build.test.ts [hunk 7]
```diff
@@ -49,6 +49,117 @@ describe("functions build", () => {
 		expect(std.err).toMatchInlineSnapshot(`""`);
 	});
 
+	it("should include any external modules imported by functions in the output bundle when the [--experimental-worker-bundle] flag is set", async () => {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages/functions-build.test.ts:52]: just moved all functions related tests to one place
